sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetricInfo.scala

```diff
@@ -27,4 +27,5 @@ import org.apache.spark.annotation.DeveloperApi
 class SQLMetricInfo(
     val name: String,
     val accumulatorId: Long,
-    val metricType: String)
+    val metricType: String,
```

Comment:
one idea is to encode the `CustomMetric` class name in the metric type: `v2_custom_com.class.name`. In the UI side we extract the class name, instantiate it, and aggregate the metrics.

We can provide some builtin implementations (Scala objects). If the class name points to a builtin implementation, do not instantiate it and simply call the method of a Scala object.

Reply:
We should also skip the metrics if we can't instantiate the class. The code may be run in the history server which doesn't have that class.

Reply:
Yes, but that's inevitable. Hopefully, most people will use the built-in `CustomMetric` implementations, so that the history server can support it as well. The worse case is people need to include the custom data source jar when starting the history server JVM.

Reply:
Okay, seems this is only feasible solution. I will update this.